cli/src/commands/workspace.rs

```diff
@@ -280,20 +281,73 @@ fn cmd_workspace_root(
     Ok(())
 }
 
+fn create_and_check_out_recovery_commit(
+    command: &CommandHelper,
+    ui: &mut Ui,
+    workspace: Workspace,
+) -> Result<Arc<ReadonlyRepo>, CommandError> {
+    let operation = command.resolve_operation(ui, workspace.repo_loader())?;
+    let repo = workspace.repo_loader().load_at(&operation)?;
+    let workspace_id = workspace.workspace_id().clone();
+    let tree_id = repo.store().empty_merged_tree_id();
+
+    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo.clone())?;
+    let (mut locked_workspace, commit) =
+        workspace_command.unchecked_start_working_copy_mutation()?;
+    let commit_id = commit.id();
+
+    let mut tx = repo.start_transaction(command.settings());
+    let mut_repo = tx.mut_repo();
+    let new_commit = mut_repo
+        .new_commit(command.settings(), vec![commit_id.clone()], tree_id.clone())
```

Comment:
Does it mean the working copy commit will be diffed against the empty tree?

I think it's preferred to be based off the known tree if the tree still exists.

Reply:
Or we can set it to `commit.tree_id().clone()`. We know that exists at least (unless the repo is corrupt, of course).

Reply:
This is a recovery flow, so I wanted something that's as problem-free and clear to the user as possible (so not trying the known tree).

If we set it to `commit.tree_id().clone()` (as Martin suggests), would we also need to `reset()` to that tree? Even if we can, I think it's better to be consistent and use the empty tree everywhere.

Reply:
I don't think the selected tree matters for rebasing; it just changes the evolution of the commit (not sure whether jj uses a different term than "evolution", but what I mean is what you see when you run `obslog`).

Let's say we want to rebase this recovery commit. As far as I know, the only trees that matter are the commit's tree and the commit's parent's tree. As far as I can tell, what we're discussing is what the commit's predecessor's tree, which is not any of the trees that matter.

Reply:
> Let's say we want to rebase this recovery commit. As far as I know, the only trees that matter are the commit's tree and the commit's parent's tree.

Ah, that's correct. So my point applies if the recovery commit is "checked out", not "edited".

Reply:
The recovery commit is indeed "checked out" during the recovery process, so that does that mean that you still want to change the tree of the recovery commit (even though it will be immediately edited to another tree)?

Reply:
> The recovery commit is indeed "checked out" during the recovery process

Is that what we want? I would expect it to be edited. Otherwise we'd just create an empty recovery commit as parent for the actual recover commit. Or what am I missing?

Reply:
Wait, the next line below here calls `set_wc_commit()`, which is what we mean by "edit", so I think we're good. I'm instead confused about what the question was, but I haven't followed this discussion closely, so feel free to ignore me.

Reply:
> So my point applies if the recovery commit is "checked out", not "edited".
Ah, if you mean "...but since the recovery commit is both checked out and edited, my point no longer applies" then we are in agreement.

Reply:
Maybe the confusion is around terminology. What I assumed Yuya meant by "checked out " is in the `jj checkout/new` sense, i.e. that we create a new working-copy commit *on top* of some commit. Conversely, "edit" means to make the working copy point to the commit itself. Note that `jj new` is a combination of "create new commit on top of specified commits" and "edit the new commit" (without `--no-edit`). Does that clarify?

Reply:
> Does it mean the working copy commit will be diffed against the empty tree?
> 
> I think it's preferred to be based off the known tree if the tree still exists.

I agree. I would expect the initial recovery commit to either be empty or contain the changes in the working copy right away. For the latter, we would need an explicit call to snapshot here after resetting the working copy. That might be a little bit more work. I'm fine with either.

So, I agree that it doesn't matter much since it's barely observable to the user, but it still seems a little less surprising if the user does inspect it if we don't include an empty tree an a predecessor.

Reply:
> What I assumed Yuya meant by "checked out " is in the `jj checkout/new` sense, i.e. that we create a new working-copy commit _on top_ of some commit.

Yes. I misinterpret that there were the recovered base commit + snapshot from then. If the snapshot is non-empty, the working directory had uncommitted changes which the user might want to rebase or cherry pick.

Reply:
OK I think we are all fine with the current implementation then? If not, let me know exactly what you think I should change and I will try it.